rllib/examples/envs/env_rendering_and_recording.py

```diff
@@ -1,135 +1,274 @@
+        self.worst_episode_and_return = (None, float("inf"))
+
+    def on_episode_step(
+        self,
+        *,
+        episode,
+        env_runner,
+        env,
+        env_index,
+        rl_module,
+        **kwargs,
+    ) -> None:
+        """On each env.step(), we add the render image to our Episode instance.
+
+        Note that this would work with MultiAgentEpisodes as well.
+        """
+        # If we have a vector env, only render the sub-env at index 0.
+        if isinstance(env.unwrapped, gym.vector.VectorEnv):
+            image = env.envs[0].render()
+        # Render the gym.Env.
+        else:
+            image = env.render()
+
+        # Original render images for CartPole are 400x600 (hxw). We'll downsize here to
+        # a very small dimension (to save space and bandwidth).
+        image = resize(image, 64, 96)
+        # For WandB videos, we need to put channels first.
+        image = np.transpose(image, axes=[2, 0, 1])
+        # Add the compiled single-step image as temp. data to our Episode object.
+        # Once the episode is done, we'll compile the video from all logged images
+        # and log the video with the EnvRunner's `MetricsLogger.log_...()` APIs.
+        # See below:
+        # `on_episode_end()`: We compile the video and maybe store it).
+        # `on_sample_end()` We log the best and worst video to the `MetricsLogger`.
+        episode.add_temporary_timestep_data("render_images", image)
+
+    def on_episode_end(
+        self,
+        *,
+        episode,
+        env_runner,
+        env,
+        env_index,
+        rl_module,
+        **kwargs,
+    ) -> None:
+        """Computes episode's return and compiles a video, iff best/worst in this iter.
+
+        Note that the actual logging to the EnvRunner's MetricsLogger only happens
+        at the very env of sampling (when we know, which episode was the best and
+        worst). See `on_sample_end` for the implemented logging logic.
         """
+        # Get the episode's return.
+        episode_return = episode.get_return()
+
+        # Better than the best or worse than worst Episode thus far?
+        if (
+            episode_return > self.best_episode_and_return[1]
+            or episode_return < self.worst_episode_and_return[1]
+        ):
+            # Pull all images from the temp. data of the episode.
+            images = episode.get_temporary_timestep_data("render_images")
 
-        # Just generate a random image here for demonstration purposes.
-        # Also see `gym/envs/classic_control/cartpole.py` for
-        # an example on how to use a Viewer object.
-        return np.random.randint(0, 256, size=(300, 400, 3), dtype=np.uint8)
+            # Create a video from the images by simply stacking them.
+            video = np.expand_dims(
+                np.stack(images, axis=0), axis=0
+            )  # TODO: test to make sure WandB properly logs videos.
```

Comment:
This TODO is important. And we should ping the Tune team to take a look into the WandB logger when using schedulers or large checkpoints.

Reply:
removed confusing TODO.